dbms/src/DataStreams/NonJoinedBlockInputStream.cpp

```diff
@@ -237,66 +270,32 @@ size_t NonJoinedBlockInputStream::fillColumns(const Map & map,
-            current_segment += step;
-            if (current_segment >= map.getSegmentSize())
-            {
-                return rows_added;
-            }
-        }
-
-        position = decltype(position)(
-            static_cast<void *>(new typename Map::SegmentType::HashTable::const_iterator(map.getSegmentTable(current_segment).begin())),
-            [](void * ptr) { delete reinterpret_cast<typename Map::SegmentType::HashTable::const_iterator *>(ptr); });
+        pos_in_hashmap = decltype(pos_in_hashmap)(
+            static_cast<void *>(new typename Map::const_iterator(map.begin())),
+            [](void * ptr) { delete reinterpret_cast<typename Map::const_iterator *>(ptr); });
+        pos_in_hashmap_inited = true;
     }
 
-    if (current_segment >= map.getSegmentSize())
-    {
-        return rows_added;
-    }
     /// use pointer instead of reference because `it` need to be re-assigned latter
-    auto it = reinterpret_cast<typename Map::SegmentType::HashTable::const_iterator *>(position.get());
-    auto end = map.getSegmentTable(current_segment).end();
+    auto it = reinterpret_cast<typename Map::const_iterator *>(pos_in_hashmap.get());
+    auto end = map.end();
 
-    for (; *it != end || current_segment + step < map.getSegmentSize();)
+    for (; *it != end;)
     {
-        if (*it == end || parent.partitions[current_segment]->isSpill())
-        {
-            current_segment += step;
-            while (parent.partitions[current_segment]->isSpill())
-            {
-                current_segment += step;
-                if (current_segment >= map.getSegmentSize())
-                {
-                    return rows_added;
-                }
-            }
-            position = decltype(position)(
-                static_cast<void *>(new typename Map::SegmentType::HashTable::const_iterator(
-                    map.getSegmentTable(current_segment).begin())),
-                [](void * ptr) { delete reinterpret_cast<typename Map::SegmentType::HashTable::const_iterator *>(ptr); });
-            it = reinterpret_cast<typename Map::SegmentType::HashTable::const_iterator *>(position.get());
-            end = map.getSegmentTable(current_segment).end();
-            continue;
-        }
-
         if ((*it)->getMapped().getUsed())
         {
             ++(*it);
             continue;
         }
 
-        rows_added += AdderNonJoined<STRICTNESS, typename Map::mapped_type>::add((*it)->getMapped(), key_num, num_columns_left, mutable_columns_left, num_columns_right, mutable_columns_right, next_element_in_row_list, max_block_size - rows_added);
-        assert(rows_added <= max_block_size);
+        rows_added += AdderNonJoined<STRICTNESS, typename Map::mapped_type>::add((*it)->getMapped(), key_num, num_columns_left, mutable_columns_left, num_columns_right, mutable_columns_right, next_element_in_row_list, max_block_size - rows_added - rows_before_fill);
+        assert(rows_added + rows_before_fill <= max_block_size);
```

Comment:
move assert before AdderNonJoined::add statement? And can add a new local variable to record "rows_added + rows_before_fill"

Reply:
Done